#399 Singlefile support

cdvv7788 opened the pull request:
# Summary

Add https://github.com/gildas-lormeau/SingleFile as an extractor

**Related issues: #399 

# Changes these areas

- [ ] Bugfixes
- [X] Feature behavior
- [ ] Command line interface
- [ ] Configuration options
- [ ] Internal architecture
- [ ] Archived data layout on disk

cdvv7788 commented:
@pirate This first version is functional but it is lacking a lot of flexibility. It uses the `CHROME_BINARY`. I need to add new options to be able to use other browsers.
Also, the installation is not possible via package managers, like it is for `wget` and `chrome`. I had to clone it, install npm dependencies and create a symlink to my `bin` folder. Should we leave this method enabled by default?

It currently works, if you have chrome installed. It creates a new file in the archive named `single-file.html`. Currently, it is created directly by the `single-file` process.

pirate commented on archivebox/config/__init__.py:
```diff
@@ -104,13 +105,15 @@
         'CURL_BINARY':              {'type': str,   'default': 'curl'},
         'GIT_BINARY':               {'type': str,   'default': 'git'},
         'WGET_BINARY':              {'type': str,   'default': 'wget'},
+        'SINGLEFILE_BINARY':        {'type': str, 'default': 'single-file'},
```
```suggestion
        'SINGLEFILE_BINARY':        {'type': str,   'default': 'single-file'},
```

pirate commented on archivebox/extractors/singlefile.py:
```diff
@@ -0,0 +1,81 @@
+    out_dir = out_dir or link.link_dir
+    output = str(Path(out_dir).absolute() / "single-file.html")
+
+    # WGET CLI Docs: https://www.gnu.org/software/wget/manual/wget.html
```
```suggestion
    # SingleFile CLI Docs: https://github.com/gildas-lormeau/SingleFile/tree/master/cli
```

pirate commented on archivebox/extractors/singlefile.py:
```diff
@@ -0,0 +1,81 @@
+
+@enforce_types
+def should_save_singlefile(link: Link, out_dir: Optional[str]=None) -> bool:
+    out_dir = out_dir or link.link_dir
```
```suggestion
    output = Path(out_dir).absolute() / "single-file.html"
```

pirate commented on archivebox/extractors/singlefile.py:
```diff
@@ -0,0 +1,81 @@
+
+@enforce_types
+def should_save_singlefile(link: Link, out_dir: Optional[str]=None) -> bool:
+    out_dir = out_dir or link.link_dir
+    if not os.path.exists(out_dir):
+        return False
+
+    return SAVE_SINGLEFILE
```
```suggestion
    output = Path(out_dir or link.link_dir) / 'single-file.html'
    return SAVE_SINGLEFILE and (not output.exists())
```

pirate commented:
A good start, things left:

- add an iframe to the link_details.html template for singlefile
- add a files column icon for SingleFile in the core.admin snapshot list
- add it to `archviebox.index.latest_outputs`
- add it to `archviebox.index.canonoical_outputs`
- anywhere else you can fine (search for `dom_path` and add it next to it everywhere)

cdvv7788 commented:
- [ ] add an iframe to the link_details.html template for singlefile
- [ ] add a files column icon for SingleFile in the core.admin snapshot list
- [ ] add it to archviebox.index.latest_outputs
- [ ] add it to archviebox.index.canonoical_outputs
- [ ] anywhere else you can fine (search for dom_path and add it next to it everywhere)

cdvv7788 commented:
@pirate please review. I think we can close this and add the new config options to the cli in another PR. Let me know if you prefer adding that in here.

pirate commented on archivebox/extractors/singlefile.py:
```diff
@@ -0,0 +1,79 @@
+    """download full site using single-file"""
+
+    out_dir = out_dir or link.link_dir
+    output = str(Path(out_dir).absolute() / "single-file.html")
```
```suggestion
    output = str(Path(out_dir).absolute() / "singlefile.html")
```

pirate commented on archivebox/extractors/singlefile.py:
```diff
@@ -0,0 +1,79 @@
+def should_save_singlefile(link: Link, out_dir: Optional[str]=None) -> bool:
+    out_dir = out_dir or link.link_dir
+
+    output = Path(out_dir or link.link_dir) / 'single-file.html'
```
```suggestion
    output = Path(out_dir or link.link_dir) / 'singlefile.html'
```

pirate commented on archivebox/index/schema.py:
```diff
@@ -365,6 +365,7 @@ def is_archived(self) -> bool:
             'screenshot.png',
             'output.html',
             'media',
+            'single-file.html'
```
```suggestion
            'singlefile.html'
```

pirate commented on archivebox/index/schema.py:
```diff
@@ -406,6 +406,7 @@ def canonical_outputs(self) -> Dict[str, Optional[str]]:
             'google_favicon_path': 'https://www.google.com/s2/favicons?domain={}'.format(self.domain),
             'wget_path': wget_output_path(self),
             'warc_path': 'warc',
+            'singlefile_path': 'single-file.html',
```
```suggestion
            'singlefile_path': 'singlefile.html',
```

pirate commented on archivebox/themes/legacy/link_details.html:
```diff
@@ -335,6 +336,18 @@ <h5>🗃 Files</h5>
+                                <img src="../../static/external.png" class="external"/>
+                            </a>
+                            <a href="$singlefile_path" target="preview"><h4 class="card-title">SingleFile</h4></a>
+                            <p class="card-text">archive/single-file.html</p>
```
```suggestion
                            <p class="card-text">archive/singlefile.html</p>
```

pirate commented on tests/test_extractors.py:
```diff
@@ -2,4 +2,10 @@
+def test_singlefile_works(tmp_path, process):
+    add_process = subprocess.run(['archivebox', 'add', 'http://127.0.0.1:8080/static/example.com.html'], capture_output=True) 
+    archived_item_path = list(tmp_path.glob('archive/**/*'))[0]
+    output_file = archived_item_path / "single-file.html" 
```
```suggestion
    output_file = archived_item_path / "singlefile.html" 
```

pirate commented:
The last thing remaining is to add it to the docker image so it works out-of-the-box.

cdvv7788 commented:
@pirate Added a fixture to the tests to avoid using methods we were not actually testing. By default it will bypass all the extractors. The python tests should be running faster now because of this change.
I also added the singlefile installation as part of the process (it was that or remove the singlefile test).
Additionally, I added a config variable that was missing (`SAVE_SINGLEFILE`) and that was causing erratic behaviour.
